### PR TITLE
allow to_gbq without datasets.create permissions

### LINKED ISSUE #73: allow to_gbq without datasets.create permissions
Currently, due to https://github.com/coiled/dask-bigquery/blob/b6d5b04e544350b2dd46e10221844d56024e92ba/dask_bigquery/core.py#L312, you cannot execute to_gbq in the situation where a dataset already exists but you do not have datasets.create permissions. Perhaps it would make sense to try `get_dataset` first, and if that doesn't work, call `create_dataset`. That would handle this situation I believe.

### PR DESCRIPTION
**Pull request by bolliger32**
Fixes #73 and supersedes #74

### CODE REVIEW

**Comment by ncclementi** [comment]
@bolliger32 Can we add a test for this, that checks for both paths? 
I'm not sure how what's the best way to approach this, maybe @j-bennet might be able to provide some guidance if needed

**Comment by j-bennet** [comment]
@ncclementi 

> Can we add a test for this, that checks for both paths? I'm not sure how what's the best way to approach this, maybe @j-bennet might be able to provide some guidance if needed

We already have tests that cover `to_gbq` without an existing dataset, we just need a test with one. It would be similar to `test_to_gbq`, but without the `write_dataset` fixture. It can still use `google_creds` fixture, but the dataset it writes to would always be the same. I created a dataset called `persistent_dataset` in our test project, it can be used for the purpose.

**Comment by ncclementi** [comment]
@bolliger32 This is almost ready, can you add the test case described by @j-bennet in the comment above?

**Comment by bolliger32** [comment]
@ncclementi I gave it a try, with my best understanding of @j-bennet 's comments. if `persistent_dataset` exists, this should test the ability to write to that existing dataset. What it won't do is text the exact functionality that I wanted to add, which is the ability to write a table to an existing dataset _when you do not have dataset.create permissions_. To do this, this might necessitate having multiple creds (with different permissions) that are tested within the test_core.py file. I might leave that up to you all as the question of whether its worth doing that is probably more a decision for maintainers.

**Comment by bolliger32** [comment]
@ncclementi I might need a little help getting this over the finish line. I’m getting errors popping up in different places with different jobs. It almost seems like it’s a simultaneity problem? Like if all these jobs are trying to read/write to the same bigquery table at the same time, that might cause some of the errors I’m seeing? Though this doesn’t seem to be occurring with the original `write_dataset` fixture, so I’m not sure what it is about the `write_existing_dataset` fixture that I added that would cause these issues. I tried to mirror the `write_dataset` structure but only delete the table, rather than the dataset, on teardown.

Also confusing is that for many of the test cases, it looks like the parameter set that caused the error initially passes but then, for some reason, is run a second time by pytest, i.e.:
image

**Comment by j-bennet** [comment]
@bolliger32 I'll take a look next week.

**Comment by bolliger32** [comment]
thanks for addressing this @j-bennet ! Just for my own understanding, why did this become an issue with the `write_existing_dataset` fixture but not the `write_dataset` fixture?

**Comment by j-bennet** [comment]
@bolliger32 `write_dataset` fixture creates and deletes a dataset with random name for every test, but `write_existing_dataset` uses the same dataset name. Within our test suite, there are multiple jobs running at the same time. If they all try to write and then delete the same table in the same dataset at the same time, they step on one another. If it's the same dataset, then at least each test instance needs its own unique table.

File: dask_bigquery/tests/test_core.py
Comment (j-bennet) [inline dask_bigquery/tests/test_core.py:230]: Very nice, love the parametrized fixture. Ingenious.

**Review by j-bennet — APPROVED** [review]
Looks good to me, and I think the test failures were fixed. Can you also take a look @ncclementi ?

**Comment by bolliger32** [comment]
> @bolliger32 `write_dataset` fixture creates and deletes a dataset with random name for every test, but `write_existing_dataset` uses the same dataset name. Within our test suite, there are multiple jobs running at the same time. If they all try to write and then delete the same table in the same dataset at the same time, they step on one another. If it's the same dataset, then at least each test instance needs its own unique table.

@j-bennet of course! makes sense

**Comment by ncclementi** [comment]
@bolliger32 and @j-bennet Thanks for your work here. I'll take a final look later today after I clear a couple of meetings. I'm seeing the mac Ci jobs timing out, but we had this problem for a while I'll ignore it.

File: dask_bigquery/tests/test_core.py
Comment (ncclementi) [inline dask_bigquery/tests/test_core.py:151]: Just for my understanding, why do we need the `None` here? Is this because in this case there is no table_id? If so can we add a comment that mentions that?

File: dask_bigquery/core.py
Comment (ncclementi) [inline dask_bigquery/core.py:323]: In the tests, I'm seeing that we are testing that we can write to an existing data set and also to create a new one. But I'm missing a test that will check that the right path is chosen based on what I read in this try/except. 

Something that checks when there is an `exception.NotFound` then the dataset will be created. Would it be possible to trigger this exception easily with the creds the bot has? @j-bennet any thoughts?

**Review by ncclementi — COMMENTED** [review]
This looks good to me, I left two comments. 

My only concern is being able to test that if we trigger the exception then we are doing the right thing. If this is easy to do we should add it, if it's too complicated we can leave it as is. 

I'll defer to @j-bennet for advice here, as I'm not super familiar with the writing part.

File: dask_bigquery/tests/test_core.py
Comment (j-bennet) [reply dask_bigquery/tests/test_core.py:151]: Yes, that's why, just so `write_existing_dataset` and `write_dataset` have the same signature.

File: dask_bigquery/core.py
Comment (j-bennet) [reply dask_bigquery/core.py:323]: We already (non-explicitly) trigger this use case whenever `write_dataset` fixture is used:

https://github.com/coiled/dask-bigquery/blob/5521666aa5f95b31c2cda6bd150c372ff529fc13/dask_bigquery/tests/test_core.py#L143-L157

Notice how the fixture does not create a dataset, it just provides you a unique id for a new dataset, and afterwards, does the cleanup.

it would still be good to test with a set of credentials for a lower-permission user though. I opened an issue for that: https://github.com/coiled/dask-bigquery/issues/76.

**Comment by ncclementi** [comment]
@j-bennet Based on your comment above, would you be ok merging this one in and we can cover the read_only credentials test separately from this PR?

**Comment by ncclementi** [comment]
@bolliger32 Thank you for your work, this goes in 🎉 
Thanks @j-bennet for reviewing!

**Pull request merged**
